submariner-operator/broker-multi-tenancy.md

```diff
@@ -0,0 +1,321 @@
+(in the case of multiple tenants in workload clusters, we could run a
+separate cluster control plane per tenant).
+- Any changes needed in runtime and data plane elements (e.g., metrics collection,
+ Lighthouse, etc.) to support configurable ClusterSet naming.
+- Support multi-tenancy in workload clusters.
+
+Note that non-goals are defined as such in the context of the currently proposed
+work and may be added in the future. They are optional for the current phase
+and may be implemented if straightforward.
+
+## Proposal
+
+The work will be carried out in stages:
+
+1. Make Broker namespace configurable (e.g. in CRDs and subctl) to
+allow multiple Broker namespaces to co-exist. The default namespace value
+can match the current fixed value to allow migration. The use of multiple
+or single namespace should be under global configuration to allow backwards
+compatibility. The change does not introduce new functionality (e.g.,
+isolation and protection) beyond configurable namespace names. This allows
+for multiple namespaces to exist concurrently, in a cooperative manner
+(i.e., discretionary access control by referencing the correct namespace).
+2. Support stricter isolation between namespaces. This step builds upon the
+previous and creates relevant service accounts, RBAC policies, etc., to ensure
+users can only access their designated Broker namespace (previous step assumed
+cooperative access).
+3. Separate Broker namespace creation and management to separate the Offering
+and ClusterSet Admin roles and privileges. This could be done via a Broker (full)
+ reconciliation loop (note: currently the broker is not active, but acts
+ as a storage namespace only). Changes should be made to minimize disruption to existing
+workflows (e.g., hide changes behind the existing `subctl` command line).
+
+The first and second steps are mostly complete by
+[submariner-operator PR 1682](https://github.com/submariner-io/submariner-operator/pull/1682).
+
+## Design Details
+
+### Overview of Proposed Changes
+
+- Add a new “central” controller/operator called Axon controller (may think
+of a better name.) to the Broker cluster. The controller runs in a new,
+singleton, Axon namespace. This controller is responsible for provisioning
+submariner tenants (i.e., a ClusterSet). The `subctl` CLI will be extended
+to support tenant management by adding the following commands (see details below):
+  - `subctl deploy axon-cp`
+  - `subctl create|delete|get|list tenant [tenant_id]`
+- Create the submariner CRDs when deploying the Axon operator, assuming that
+tenants (i.e., ClusterSet Admins) will not have permissions to create those.
+The `deploy broker` command will ensure the CRDs exist.
+- A submariner tenant is represented by a tenant object in the Axon namespace
+ and a dedicated tenant namespace. Each tenant has a string ID and the namespace
+ is some constant prefix + tenant ID.
+- A SA is created per tenant and a *partial* `broker-info.subm` file is generated
```

Comment:
So why not create a `broker-info.subm.template` or something?

Reply:
So that user can already use the generated `subm` file with `subctl` to connect to the to the broker cluster with the correct user and namespace context